[clang-tidy] add modernize-use-constexpr check

5chmidti opened the pull request:
PR will be split for easier review

This check finds all functions and variables that can be declared as
`constexpr`, using the specified standard version to check if the
requirements are met.

Fixes #115622

5chmidti committed:
[clang-tidy] add misc-constexpr check

This check finds all functions and variables that can be declared as
`constexpr`, using the specified standard version to check if the
requirements are met.

Fixes #115622

carlosgalvezp commented:
misc-use-constexpr? :)

5chmidti commented:
I ran this check on all of `clang/` and `clang-tools-extra/` and got no compilation errors from too aggressively changing things to `constexpr`.

Some perf info from running the checks enabled in LLVM + this check:

Running on: `clang/lib/Sema/SemaExprCXX.cpp` (1 diag):
```text
===-------------------------------------------------------------------------===
                          clang-tidy checks profiling
===-------------------------------------------------------------------------===
  Total Execution Time: 6.4289 seconds (6.4453 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
   0.7024 ( 20.6%)   0.7401 ( 24.5%)   1.4424 ( 22.4%)   1.4457 ( 22.4%)  misc-unused-using-decls
   0.6188 ( 18.2%)   0.1852 (  6.1%)   0.8040 ( 12.5%)   0.8069 ( 12.5%)  llvm-qualified-auto
   0.2833 (  8.3%)   0.2724 (  9.0%)   0.5556 (  8.6%)   0.5581 (  8.7%)  llvm-prefer-isa-or-dyn-cast-in-conditionals
   0.2288 (  6.7%)   0.2426 (  8.0%)   0.4714 (  7.3%)   0.4715 (  7.3%)  misc-misleading-identifier
   0.2271 (  6.7%)   0.2140 (  7.1%)   0.4411 (  6.9%)   0.4423 (  6.9%)  misc-confusable-identifiers
   0.2132 (  6.3%)   0.2268 (  7.5%)   0.4400 (  6.8%)   0.4402 (  6.8%)  misc-definitions-in-headers
   0.1938 (  5.7%)   0.2066 (  6.8%)   0.4003 (  6.2%)   0.4008 (  6.2%)  misc-non-copyable-objects
   0.1371 (  4.0%)   0.1386 (  4.6%)   0.2757 (  4.3%)   0.2750 (  4.3%)  misc-unconventional-assign-operator
   0.1235 (  3.6%)   0.1296 (  4.3%)   0.2531 (  3.9%)   0.2549 (  4.0%)  misc-constexpr
   0.1146 (  3.4%)   0.1188 (  3.9%)   0.2334 (  3.6%)   0.2346 (  3.6%)  misc-misplaced-const
   0.1176 (  3.5%)   0.1061 (  3.5%)   0.2238 (  3.5%)   0.2236 (  3.5%)  misc-redundant-expression
   0.1074 (  3.2%)   0.1110 (  3.7%)   0.2184 (  3.4%)   0.2208 (  3.4%)  misc-use-internal-linkage
   0.0665 (  2.0%)   0.0720 (  2.4%)   0.1385 (  2.2%)   0.1400 (  2.2%)  misc-new-delete-overloads
   0.0615 (  1.8%)   0.0640 (  2.1%)   0.1255 (  2.0%)   0.1257 (  2.0%)  llvm-prefer-register-over-unsigned
   0.0548 (  1.6%)   0.0572 (  1.9%)   0.1120 (  1.7%)   0.1121 (  1.7%)  llvm-twine-local
   0.0444 (  1.3%)   0.0328 (  1.1%)   0.0771 (  1.2%)   0.0773 (  1.2%)  llvm-namespace-comment
   0.0385 (  1.1%)   0.0312 (  1.0%)   0.0697 (  1.1%)   0.0695 (  1.1%)  llvm-else-after-return
   0.0325 (  1.0%)   0.0351 (  1.2%)   0.0677 (  1.1%)   0.0674 (  1.0%)  misc-unused-alias-decls
   0.0294 (  0.9%)   0.0255 (  0.8%)   0.0549 (  0.9%)   0.0547 (  0.8%)  misc-static-assert
   0.0108 (  0.3%)   0.0100 (  0.3%)   0.0208 (  0.3%)   0.0207 (  0.3%)  misc-uniqueptr-reset-release
   0.0018 (  0.1%)   0.0018 (  0.1%)   0.0036 (  0.1%)   0.0035 (  0.1%)  misc-misleading-bidirectional
   0.0000 (  0.0%)   0.0000 (  0.0%)   0.0000 (  0.0%)   0.0000 (  0.0%)  misc-throw-by-value-catch-by-reference
   3.4077 (100.0%)   3.0212 (100.0%)   6.4289 (100.0%)   6.4453 (100.0%)  Total
```

   
Running on `clang/utils/TableGen/RISCVVEmitter.cpp` (3 diags):
```text
===-------------------------------------------------------------------------===
                          clang-tidy checks profiling
===-------------------------------------------------------------------------===
  Total Execution Time: 1.1712 seconds (1.1748 wall clock)

   ---User Time---   --System Time--   --User+System--   ---Wall Time---  --- Name ---
   0.1293 ( 20.2%)   0.1265 ( 23.8%)   0.2557 ( 21.8%)   0.2569 ( 21.9%)  misc-unused-using-decls
   0.0673 ( 10.5%)   0.0576 ( 10.9%)   0.1249 ( 10.7%)   0.1257 ( 10.7%)  llvm-prefer-isa-or-dyn-cast-in-conditionals
   0.0833 ( 13.0%)   0.0275 (  5.2%)   0.1108 (  9.5%)   0.1112 (  9.5%)  llvm-qualified-auto
   0.0434 (  6.8%)   0.0407 (  7.7%)   0.0840 (  7.2%)   0.0843 (  7.2%)  misc-misleading-identifier
   0.0436 (  6.8%)   0.0385 (  7.2%)   0.0820 (  7.0%)   0.0817 (  7.0%)  misc-confusable-identifiers
   0.0417 (  6.5%)   0.0387 (  7.3%)   0.0804 (  6.9%)   0.0806 (  6.9%)  misc-definitions-in-headers
   0.0383 (  6.0%)   0.0359 (  6.8%)   0.0742 (  6.3%)   0.0744 (  6.3%)  misc-non-copyable-objects
   0.0275 (  4.3%)   0.0217 (  4.1%)   0.0492 (  4.2%)   0.0490 (  4.2%)  misc-redundant-expression
   0.0229 (  3.6%)   0.0215 (  4.1%)   0.0445 (  3.8%)   0.0443 (  3.8%)  misc-constexpr
   0.0229 (  3.6%)   0.0203 (  3.8%)   0.0432 (  3.7%)   0.0431 (  3.7%)  misc-unconventional-assign-operator
   0.0200 (  3.1%)   0.0191 (  3.6%)   0.0391 (  3.3%)   0.0391 (  3.3%)  misc-misplaced-const
   0.0199 (  3.1%)   0.0185 (  3.5%)   0.0384 (  3.3%)   0.0390 (  3.3%)  misc-use-internal-linkage
   0.0154 (  2.4%)   0.0081 (  1.5%)   0.0235 (  2.0%)   0.0236 (  2.0%)  llvm-namespace-comment
   0.0114 (  1.8%)   0.0111 (  2.1%)   0.0225 (  1.9%)   0.0229 (  1.9%)  llvm-prefer-register-over-unsigned
   0.0115 (  1.8%)   0.0104 (  2.0%)   0.0219 (  1.9%)   0.0222 (  1.9%)  misc-new-delete-overloads
   0.0104 (  1.6%)   0.0102 (  1.9%)   0.0206 (  1.8%)   0.0207 (  1.8%)  llvm-twine-local
   0.0122 (  1.9%)   0.0083 (  1.6%)   0.0205 (  1.8%)   0.0205 (  1.7%)  misc-static-assert
   0.0108 (  1.7%)   0.0074 (  1.4%)   0.0182 (  1.6%)   0.0181 (  1.5%)  llvm-else-after-return
   0.0069 (  1.1%)   0.0067 (  1.3%)   0.0135 (  1.2%)   0.0135 (  1.2%)  misc-unused-alias-decls
   0.0013 (  0.2%)   0.0012 (  0.2%)   0.0025 (  0.2%)   0.0025 (  0.2%)  misc-uniqueptr-reset-release
   0.0006 (  0.1%)   0.0008 (  0.1%)   0.0014 (  0.1%)   0.0014 (  0.1%)  misc-misleading-bidirectional
   0.0000 (  0.0%)   0.0000 (  0.0%)   0.0000 (  0.0%)   0.0000 (  0.0%)  misc-throw-by-value-catch-by-reference
   0.6406 (100.0%)   0.5306 (100.0%)   1.1712 (100.0%)   1.1748 (100.0%)  Total
```

denzor200 commented:
> misc-use-constexpr? :)

modernize-use-constexpr?

vbvictor commented on clang-tools-extra/clang-tidy/misc/CMakeLists.txt:
```diff
@@ -19,6 +19,7 @@ set_target_properties(genconfusable PROPERTIES FOLDER "Clang Tools Extra/Sourceg
 
 add_clang_library(clangTidyMiscModule STATIC
   ConstCorrectnessCheck.cpp
+  UseConstexprCheck.cpp
```
Please fix ordering

vbvictor commented on clang-tools-extra/clang-tidy/misc/MiscTidyModule.cpp:
```diff
@@ -43,6 +44,7 @@ class MiscModule : public ClangTidyModule {
         "misc-confusable-identifiers");
     CheckFactories.registerCheck<ConstCorrectnessCheck>(
         "misc-const-correctness");
+    CheckFactories.registerCheck<UseConstexprCheck>("misc-use-constexpr");
```
ditto

vbvictor commented on clang-tools-extra/clang-tidy/misc/UseConstexprCheck.cpp:
```diff
@@ -0,0 +1,946 @@
+  return Node.isConstexprSpecified();
+}
+
+bool satisfiesConstructorPropertiesUntil20(const CXXConstructorDecl *Ctor,
```
Regular functions should be declared outside of anon-namespace.
Soon my new check will flag this:)
Same applied to all other functions

vbvictor commented on clang-tools-extra/clang-tidy/misc/UseConstexprCheck.cpp:
```diff
@@ -0,0 +1,946 @@
+                                           ASTContext &Ctx) {
+  const CXXRecordDecl *Rec = Ctor->getParent();
+  llvm::SmallPtrSet<const RecordDecl *, 8> Bases{};
+  for (const CXXBaseSpecifier Base : Rec->bases()) {
+    Bases.insert(Base.getType()->getAsRecordDecl());
+  }
```
```suggestion
  for (const CXXBaseSpecifier Base : Rec->bases())
    Bases.insert(Base.getType()->getAsRecordDecl());
```

vbvictor commented on clang-tools-extra/clang-tidy/misc/UseConstexprCheck.cpp:
```diff
@@ -0,0 +1,946 @@
+    const FunctionDecl *FDecl, ASTContext &Ctx,
+    const bool ConservativeLiteralType,
+    const bool AddConstexprToMethodOfClassWithoutConstexprConstructor) {
+  if (FDecl->isConstexprSpecified()) {
+    return true;
+  }
```
```suggestion
  if (FDecl->isConstexprSpecified())
    return true;
```

vbvictor commented on clang-tools-extra/clang-tidy/misc/UseConstexprCheck.cpp:
```diff
@@ -0,0 +1,946 @@
+    const FunctionDecl *FDecl, ASTContext &Ctx,
+    const bool ConservativeLiteralType,
+    const bool AddConstexprToMethodOfClassWithoutConstexprConstructor) {
+  if (FDecl->isConstexprSpecified()) {
+    return true;
+  }
```
```suggestion
  if (FDecl->isConstexprSpecified())
    return true;
```

vbvictor commented on clang-tools-extra/clang-tidy/misc/UseConstexprCheck.cpp:
```diff
@@ -0,0 +1,946 @@
+bool satisfiesProperties2326(
+    const FunctionDecl *FDecl, ASTContext &Ctx,
+    const bool AddConstexprToMethodOfClassWithoutConstexprConstructor) {
+  if (FDecl->isConstexprSpecified()) {
+    return true;
+  }
```
```suggestion
  if (FDecl->isConstexprSpecified())
    return true;
```

denzor200 commented on clang-tools-extra/test/clang-tidy/checkers/misc/use-constexpr.cpp:
```diff
@@ -0,0 +1,562 @@
+        // CHECK-MESSAGES-11: :[[@LINE-1]]:13: warning: function 'g1' can be declared 'constexpr' [misc-use-constexpr]
+        // CHECK-FIXES-11: constexpr int g1() { return 0; }
+        static constexpr int A1 = 0;
+        static int B1 = 0;
```
Just asking - this check will never change it to `static constexpr int B1 = 0;`, is that intended?

denzor200 replied:
As I understand, this is a job for "misc-const-correctness", and then this check should be runned.
Thus we should write in the doc a recomendation to run this check only after "misc-const-correctness" have done.

5chmidti replied:
I will extend the check docs to explain the dependency on `const` for variables.

5chmidti committed:
rename to modernize-use-constexpr + review comments

5chmidti committed:
fix formatting

vbvictor commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -270,6 +270,10 @@ Changes in existing checks
   excluding variables with ``thread_local`` storage class specifier from being
   matched.
 
+- Added :doc:`modernize-use-constexpr
```
This should be placed under "New checks" section

vbvictor replied:
Also, synced with first sentence in docs.

schenker commented:
I got some fixits that do not compile, e.g.:
```
std::erase_if(numbers, [](int n) {
return n % 2 == 0;
});
```
is changed into
```
std::erase_if(numbers, [](int nconstexpr ) {
return n % 2 == 0;
});
```

5chmidti committed:
add ConstexprString option

5chmidti committed:
align short explanation of check

5chmidti committed:
move release note entry to correct location

5chmidti committed:
fix removal range of `const` kw

5chmidti committed:
move replacement string creation out of loop

5chmidti commented:
> I tried with [ITK](https://github.com/InsightSoftwareConsortium/ITK/) and the resulting transformation did not compile.
> 
> Many changes were of this form:
> 
> ```diff
> -    const double     Max = 1.0 - Min;
> +    constexpr  double     Max = 1.0 - Min;
> ```
> 
> Which is great, though notice the double space after `constexpr` for some reason.

The reason is that when inserting, I add a `constexpr ` with a trailing space, but when it is found, also remove the `const` token without its corresponding space. This will have to be fixed by running clang-format; we have a few of these additional whitespaces in some checks.

> Other changes were like this:
> 
> ```diff
> -  const auto check = [](const auto & ptr) { EXPECT_THROW(itk::Deref(ptr), itk::DerefError); };
> +  constexpr const auto check = [](auto & ptr) { EXPECT_THROW(itk::Deref(ptr), itk::DerefError); };
> ```
> 
> I'm no C++ expert, but is it right to have both `const` _**and**_ `constexpr` here?
> 
> Also, I was surprised to see the 2nd const _removed_. And this removal generates one of the many compiler errors.

Eh... yeah. The range I passed to the `getQualifyingToken` function was too big and included more than it needed. It's fixed now, and I've added a test for it.

> Still, this is looking like it'll be great!

Thanks, and thank you very much for trying it out. My current setup is limited in what I can test outside the LLVM codebase.

5chmidti commented:
Thanks for testing this on your codebase so fast as well.

> I got some fixits that do not compile, e.g.:
> 
> ```
> std::erase_if(numbers, [](int n) {
> return n % 2 == 0;
> });
> ```
> 
> is changed into
> 
> ```
> std::erase_if(numbers, [](int nconstexpr ) {
> return n % 2 == 0;
> });
> ```


I've tried to reproduce this but was unable to, could you please provide a small reproducer?
I've tried this, and it did not produce a diagnostic (using C++20):

```c++
#include <vector>

int main() {
  std::vector<int> Numbers = {
      0, 1, 2, 3, 4, 5, 6, 7,
  };
  const auto count = std::erase_if(Numbers, [](int N) { return N % 2 == 0; });
}
```

5chmidti committed:
add StaticConstexprString option

5chmidti committed:
explain `const` dependency in docs + small docs fixes

5chmidti commented:
Sorry for the silly docs mistakes, btw. 

I've updated the docs, added replacement string options, and fixed the `const` keyword removal.

5chmidti committed:
fix doc inline code line break

5chmidti committed:
fix docs links and literal

schenker commented:
> I've tried to reproduce this but was unable to, could you please provide a small reproducer?

With 127d7f00847e6f53af75534187cd70878e635636 I can reporoduce the wrong fixit like this:

**test.cpp**
```cpp
#include <vector>
template <typename T> void func() {
  std::vector<int> Numbers = {0, 1};
  const auto count =
  std::erase_if(Numbers, [](int N) { return N % 2 == 0; });
}
```
**commands**
```console
clang-tidy -checks=-*,*constexpr* -export-fixes fixes.yaml test.cpp -- --std=c++23
clang-apply-replacements .
```
**result**
```cpp
#include <vector>
template <typename T> void func() {
  std::vector<int> Numbers = {0, 1};
  const auto count =
  std::erase_if(Numbers, [](int Nconstexpr ) { return N % 2 == 0; });
}
```

EugeneZelenko commented on clang-tools-extra/clang-tidy/modernize/UseConstexprCheck.cpp:
```diff
@@ -0,0 +1,985 @@
+    return true;
+
+  const LangOptions LO = Ctx.getLangOpts();
+  const CXXMethodDecl *Method = llvm::dyn_cast<CXXMethodDecl>(FDecl);
```
```suggestion
  const auto *Method = llvm::dyn_cast<CXXMethodDecl>(FDecl);
```

EugeneZelenko commented on clang-tools-extra/clang-tidy/modernize/UseConstexprCheck.cpp:
```diff
@@ -0,0 +1,985 @@
+  if (!Finder->getASTContext().getLangOpts().CPlusPlus20)
+    return true;
+
+  const auto *RDecl = T->getAsCXXRecordDecl();
```
Type, please. Same below.

EugeneZelenko commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -142,6 +142,10 @@ New checks
   Finds unscoped (non-class) ``enum`` declarations and suggests using
   ``enum class`` instead.
 
+- New :doc:`modernize-use-constexpr
+  <clang-tidy/checks/modernize/use-constexpr>` check that finds functions and
+  variables that can be declared `constexpr`.
```
```suggestion
- New :doc:`modernize-use-constexpr
  <clang-tidy/checks/modernize/use-constexpr>` check.
  
   Finds functions and  variables that can be declared ``constexpr``.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-constexpr.rst:
```diff
@@ -0,0 +1,64 @@
+modernize-use-constexpr
+=======================
+
+Finds functions and variables that can be declared 'constexpr'.
```
```suggestion
Finds functions and variables that can be declared '`constexpr'`.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-constexpr.rst:
```diff
@@ -0,0 +1,64 @@
+  While a function of a class or struct could be declared ``constexpr``, when
+  the class itself can never be constructed at compile-time, then adding
+  ``constexpr`` to a member function is superfluous. This option controls if
+  ``constexpr`` should be added anyways. Default is ``false``.
```
```suggestion
  ``constexpr`` should be added anyways. Default is `false`.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-constexpr.rst:
```diff
@@ -0,0 +1,64 @@
+.. option:: ConstexprString
+
+  The string to use to specify a variable or function as ``constexpr``, for
+  example, a macro. Default is ``constexpr``.
```
```suggestion
  example, a macro. Default is `constexpr`.
```

EugeneZelenko commented on clang-tools-extra/docs/clang-tidy/checks/modernize/use-constexpr.rst:
```diff
@@ -0,0 +1,64 @@
+.. option:: ConstexprString
+
+  The string to use with C++23 to specify a function-local variable as 
+  ``static constexpr``, for example, a macro. Default is ``static constexpr``
```
```suggestion
  ``static constexpr``, for example, a macro. Default is `static constexpr`
```

seanm commented:
Another compiler error after transformation:

It made this constexpr:

```c++
	PUGI_IMPL_FN constexpr bool is_nan(double value)
	{
	#if defined(PUGI_IMPL_MSVC_CRT_VERSION) || defined(__BORLANDC__)
		return !!_isnan(value);
	#elif defined(fpclassify) && defined(FP_NAN)
		return fpclassify(value) == FP_NAN;
	#else
		// fallback
		const volatile double v = value;
		return v != v;
	#endif
	}
```

But then:

```
/Users/sean/external/VTK/ThirdParty/pugixml/vtkpugixml/src/pugixml.cpp:8413:30: error: constexpr function never produces a constant expression [-Winvalid-constexpr]
 8413 |         PUGI_IMPL_FN constexpr bool is_nan(double value)
      |                                     ^~~~~~
/Users/sean/external/VTK/ThirdParty/pugixml/vtkpugixml/src/pugixml.cpp:8422:15: note: read of volatile-qualified type 'const volatile double' is not allowed in a constant expression
 8422 |                 return v != v;
      |                             ^
```

5chmidti commented:
> Another interesting case, it seems somehow multiple constexprs got added, ex:
> 
> ```c++
> template <typename CellType>
> constexpr constexpr constexpr constexpr constexpr constexpr constexpr constexpr int numberOfSidesOfDimension(int dimension);
> ```
> 
> This actually compiles, though with a `-Wduplicate-decl-specifier` warning.

I failed to reproduce this with

```c++
template <typename T>
static T forwardDeclared();

template <typename T>
static T forwardDeclared() { return T{}; }

static void useForwardDeclared() {
    forwardDeclared<int>() + forwardDeclared<double>() + forwardDeclared<char>();
}
```

```diff
 template <typename T>                                                                                                                                        
-static T forwardDeclared();                                
+constexpr static T forwardDeclared();                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
 template <typename T>                                                      
-static T forwardDeclared() { return T{}; }
+constexpr static T forwardDeclared() { return T{}; }                                                                                                                                                                                                                                                                       
                                                                  
 static void useForwardDeclared() {                                                                                                                                                                                                                                                                                         
     forwardDeclared<int>() + forwardDeclared<double>() + forwardDeclared<char>();  
```

I had seen this one as well, but it was already fixed in the initial commit, at least I thought.

seanm commented:
> > This actually compiles, though with a `-Wduplicate-decl-specifier` warning.
> 
> I failed to reproduce this...

I don't have a minimal repro case, I was building [VTK](https://gitlab.kitware.com/vtk/vtk). It's pretty simple to build it yourself if you're familiar with CMake.  I could give you exact command line steps if you need...

seanm commented:
I've tried another project: [opencv](https://github.com/opencv/opencv) and the result also fails to compile.

One case I think is because two variables are declared together. i.e.:

```
int foo, bar;
```

instead of:

```
int foo;
int bar;
```

Concretely:

```c++
    const char * const borderMap[] = { "BORDER_CONSTANT", "BORDER_REPLICATE", "BORDER_REFLECT", "BORDER_WRAP", "BORDER_REFLECT_101" },
        * const btype = borderMap[borderType & ~BORDER_ISOLATED];
```

Got changed to:

```
    constexpr const char * borderMap[] = { "BORDER_CONSTANT", "BORDER_REPLICATE", "BORDER_REFLECT", "BORDER_WRAP", "BORDER_REFLECT_101" },
        * const btype = borderMap[borderType & ~BORDER_ISOLATED];
```

which is fine for `borderMap` but not for `btype`.

Resulting in:

```
/Users/sean/external/opencv/modules/imgproc/src/filter.dispatch.cpp:757:17: error: constexpr variable 'btype' must be initialized by a constant expression
  757 |         * const btype = borderMap[borderType & ~BORDER_ISOLATED];
      |                 ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

5chmidti committed:
fix a few issues

5chmidti committed:
fix lambda call operator matching

5chmidti committed:
fix remaining auto usage issues

5chmidti committed:
fix docs

5chmidti committed:
remove unused function

5chmidti committed:
exclude lambda operators earlier

5chmidti committed:
fix constexpr insertion location

The insertion location was broken due to the insertion point being
possibly in front of attributes

vbvictor reviewed (commented):
Gave this a very rought look, so only a couple of general questions/suggestions.

Could we do something with methods that repeat in some of the visitors, e.g. `bool TraverseType(QualType QT)` and others. Maybe macros..? I can't think of a better idea right now.

Could we break this PR in multiple ones, e.g. first with support of CXX11, then CXX14 and so on? 1000-line `.cpp` file is hard to review. Also, with a PR for each standart we will understand better what changed and what new constexpr cases appear.

5chmidti committed:
fix multiple variables declared in one statement

5chmidti committed:
add newline

5chmidti committed:
fix potential issue in unrwapPointee

5chmidti commented:
Fixed multiple vars declared in one statement

5chmidti commented:
> Gave this a very rought look, so only a couple of general questions/suggestions.
> 
> Could we do something with methods that repeat in some of the visitors, e.g. `bool TraverseType(QualType QT)` and others. Maybe macros..? I can't think of a better idea right now.

I will try to think of something

> Could we break this PR in multiple ones, e.g. first with support of CXX11, then CXX14 and so on? 1000-line `.cpp` file is hard to review. Also, with a PR for each standart we will understand better what changed and what new constexpr cases appear.

That's a valid point. I'll keep this as is for a while longer without the expectation that people will review it, so that I can resolve a few issues. I also want to fix my setup to be able to run on external code bases. Then I will split it up into {[11],[14, 17],[20],[23, 26]} for 4 reviews. I'm not splitting per version since the two pairs of versions are quite similar to each other. The PRs will probably be serialized and not stacked, because we all know how nice GitHubs PR stacking works.

EugeneZelenko commented:
Please rebase from `main`. Release Notes were cleared after 21 branch.

seanm commented:
> Then I will split it up into {[11],[14, 17],[20],[23, 26]}

BTW does this support C23?  That would be nice to have too.

5chmidti commented:
> > Then I will split it up into {[11],[14, 17],[20],[23, 26]}
> 
> BTW does this support C23?  That would be nice to have too.

I totally forgot to add that :) I'll take a look at the rules for it, but might only add it later

5chmidti commented:
I'll have very little access to my machine this month, but I'll move this forward in October (maybe with a bit before that) and also split the PR into smaller ones.
